### PR TITLE
Pippa/pe visualisations

### PR DESCRIPTION
**Pull request by pippaduckett1**
add PE logging + reorganise addition of hit and query PE and sorting in phi - broke with merge main, now working but with old task.py - looking into what has gone wrong (left older task.py here but can remove)

### CODE REVIEW

File: src/hepattn/models/decoder.py [hunk 1]
```diff
@@ -123,50 +153,147 @@ def forward(self, x: dict[str, Tensor], input_names: list[str]) -> tuple[dict[st
                 for input_name, task_attn_mask in attn_masks.items():
                     attn_mask[..., x[f"key_is_{input_name}"]] = task_attn_mask
 
+            # do sorting
+            # TODO: would like to move this outside of layers for loop but slightly more complicated in terms of what to undo when
+            # - would need to sort and unsort more vars and easy to make errors or miss something? could change?
+            # TODO: would also be nice if we sorted before the encoder and then unsorted after producing decoder layer?
```
Comment (samvanstroud) [inline src/hepattn/models/decoder.py:159]: I think best to avoid the unsort in the encoder, or even move the initial sort to a dedicated step before the encoder with some of the other boilerplate, and make it optional to unsort things there. cc @trahxam

File: src/hepattn/models/decoder.py [hunk 2]
```diff
@@ -123,50 +153,147 @@ def forward(self, x: dict[str, Tensor], input_names: list[str]) -> tuple[dict[st
...
+    def get_sort_idx(self, phi: Tensor) -> Tensor:
+        if len(phi.shape) == 2:
+            phi = phi[0]
+        assert len(phi.shape) == 1, "Phi must be 1D"
+        return torch.argsort(phi)
+
+    def get_unsort_idx(self, sort_idx: Tensor) -> Tensor:
+        if len(sort_idx.shape) == 2:
+            sort_idx = sort_idx[0]
+        assert len(sort_idx.shape) == 1, "Sort index must be 1D"
```
Comment (samvanstroud) [inline src/hepattn/models/decoder.py:295]: feels like we should be able to remove all of this once the sorting logic is moved elsewhere

File: src/hepattn/models/task.py [hunk 3]
```diff
@@ -28,10 +28,9 @@ class Task(nn.Module, ABC):
     that can be trained as part of a multi-task learning setup.
     """
```
Comment (samvanstroud) [inline src/hepattn/models/task.py:30]: changes in this module should be reverted I think?

**Review by samvanstroud — COMMENTED** [review]
Thanks @pippaduckett1! I left a few initial comments. Mainly I think a lot of the PE diagnostics should be moved out of a callback and into more modular functions that could be called from e.g. tests or simple scripts which search the PE parameter space, since determining the right settings for the PE doesn't require training a model.

File: src/hepattn/models/decoder.py (hunk 1 above)
Comment (pippaduckett1) [reply src/hepattn/models/decoder.py:159]: Yeah this makes sense I agree - I had initially avoided this because I was scared I'd miss something that needed to be sorted - but as far as I can see all variables that need to be sorted would be:
`inputs[raw_var]`
`inputs[input_name+"_valid"]`

Then in the loss would we need to sort `targets[target_object+"_valid"]`?
Or better to unsort the outputs at the end of forward()?

File: src/hepattn/models/task.py (hunk 3 above)
Comment (pippaduckett1) [reply src/hepattn/models/task.py:30]: Yes - I left this in because when I was trying to work out the drop in performance after merging main, reverting `task.py` regained performance - I haven't yet isolated exactly which change to `task.py` led to the reduction in peformance so until then I'll revert changes

File: src/hepattn/models/decoder.py (hunk 1 above)
Comment (pippaduckett1) [reply src/hepattn/models/decoder.py:144]: Yeah I think this would be similar - in that case, is there anywhere that this could be a problem?
Eg. does the order of `preds` matter if we call `predict()` or would the order matter if `forward()` was called elsewhere (ie. not in `loss()`)?

**Comment by pippaduckett1** [comment]
Thanks so much for the comments - should be ready to merge as long as you're happy with the way sorting is now done in the loss :)

File: src/hepattn/utils/sorting.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
+import torch
+from torch import Tensor, nn
+
+
+class Sorter:
```
Comment (samvanstroud) [inline src/hepattn/utils/sorting.py:5]: If this has modules as members I think it should probably inherit from `nn.Module` itself

File: src/hepattn/models/decoder.py [hunk 5]
```diff
@@ -68,42 +76,52 @@ def forward(self, x: dict[str, Tensor], input_names: list[str]) -> tuple[dict[st
...
-                        attn_masks[input_name] = attn_mask
+            attn_masks: dict[str, torch.Tensor] = {}
+            query_mask = None
 
-                # Collect query masks
-                if self.use_query_masks:
-                    task_query_mask = task.query_mask(task_outputs)
-                    if task_query_mask is not None:
-                        query_mask = task_query_mask if query_mask is None else query_mask | task_query_mask
+            if self.tasks is not None:
```
Comment (samvanstroud) [inline src/hepattn/models/decoder.py:95]: `self.tasks` should never be `None` at this stage, so I would remove this `if` to reduce indentation and have an explicit error on the next line if something has gone wrong

File: src/hepattn/utils/sorting.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+        Returns:
+        -------
+        dict[str, Tensor]
+            Sort indices for key and query dimensions.
+        """
+        self.sort_indices = {}
+
+        # Get key_embed shape for reference in sorting
+        self.num_hits = self._get_num_hits(x)
+        assert self.num_hits > 0, "key embed not found in x"
```
Comment (samvanstroud) [inline src/hepattn/utils/sorting.py:33]: I would remove the `_get_num_hits()`, it's only used once. We can just do

```
self.num_hits = x["key_embed"].shape[-2]
```

Which will nautrally raise an `KeyError` if the expected inputs are not present

File: src/hepattn/utils/sorting.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+        key_sort_value = x[f"key_{self.input_sort_field}"]
+        key_sort_idx = torch.argsort(key_sort_value, dim=-1)
+        if len(key_sort_idx.shape) == 2:
+            key_sort_idx = key_sort_idx[0]
+        assert len(key_sort_idx.shape) == 1, "Sort index must be 1D"
+        assert key_sort_idx.shape[0] == self.num_hits, f"Key sort index shape {key_sort_idx.shape} does not match num_hits {self.num_hits}"
+        self.sort_indices["key"] = key_sort_idx
+
+        # TODO: sort key_phi (key_{input_sort_field})
```
Comment (samvanstroud) [inline src/hepattn/utils/sorting.py]: Is this still todo?

File: src/hepattn/utils/sorting.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+            if raw_var in x and x[raw_var].shape[-1] == x["key_embed"].shape[-2]:
+                x[raw_var] = self._sort_tensor_by_index(x[raw_var], key_sort_idx, self.num_hits)
+            else:
+                print(f"Warning: Raw variable {raw_var} has invalid shape: {x[raw_var].shape}")
+
+        return x
+
+    def sort_targets(self, targets: dict, sort_indices: dict[str, Tensor]) -> dict:
+        """Sort targets to align with sorted outputs."""
+        # TODO: check that this sorts all the targets correctly
```
Comment (samvanstroud) [inline src/hepattn/utils/sorting.py]: Is this still todo? In particular do we need to still sort the targets which are unmerged?

File: src/hepattn/utils/sorting.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+            Sort indices.
+        num_hits : int
+            Number of hits.
+
+        Returns:
+        Tensor
+            Sorted tensor.
+        """
+        if tensor is None:
+            return None
```
Comment (samvanstroud) [inline src/hepattn/utils/sorting.py:98]: Is this an expected condition?

File: src/hepattn/utils/sorting.py [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+        sort_dim = None
+        # If num hits is provided, find the dimension with matching size
+        for dim, size in enumerate(tensor.shape):
+            if size == num_hits:
+                sort_dim = dim
+                break
+
+        if sort_dim is not None:
+            return tensor.index_select(sort_dim, sort_idx.to(tensor.device))
```
Comment (samvanstroud) [inline src/hepattn/utils/sorting.py:107]: Let's assume the tensor is already on the correct device and remove `.to(tensor.device)`

File: src/hepattn/models/wrapper.py [hunk 11]
```diff
@@ -79,10 +79,9 @@ def training_step(self, batch, batch_idx):
         inputs, targets = batch
 
         # Get the model outputs
-        outputs = self.model(inputs)
-
+        outputs, sort_indices = self.model(inputs)
```
Comment (samvanstroud) [inline src/hepattn/models/wrapper.py:82]: any reason not to put `sort_indices` as a item in the `outputs` dict?

File: src/hepattn/models/maskformer.py [hunk 12]
```diff
@@ -234,12 +273,12 @@ def loss(self, outputs: dict, targets: dict) -> dict:
                     outputs[layer_name][task.name][output_name] = outputs[layer_name][task.name][output_name][batch_idxs, pred_idxs]
 
         # Compute the losses for each task in each block
-        losses = {}
+        losses: dict[str, dict[str, Any]] = {}
```
Comment (samvanstroud) [inline src/hepattn/models/maskformer.py:276]: I think we can replace `Any` with `Tensor`? probably better to be specific if we are adding type hints

File: src/hepattn/models/maskformer.py [hunk 13]
```diff
@@ -219,7 +255,10 @@ def loss(self, outputs: dict, targets: dict) -> dict:
                 continue
 
             # Get the indicies that can permute the predictions to yield their optimal matching
-            pred_idxs = self.matcher(cost, targets[f"{self.target_object}_valid"])
+            if self.matcher is not None:
+                pred_idxs = self.matcher(cost, targets[f"{self.target_object}_valid"])
+            else:
+                continue
```
Comment (samvanstroud) [inline src/hepattn/models/maskformer.py:261]: Why this change? Don't we always have a matcher?

File: src/hepattn/models/maskformer.py [hunk 14]
```diff
@@ -111,13 +117,22 @@ def forward(self, inputs: dict[str, Tensor]) -> dict[str, Tensor]:
                 [inputs[input_name + "_" + self.input_sort_field] for input_name in input_names], dim=-1
             )
 
+        # Dedicated sorting step before encoder
+        self.sorting = Sorter(input_sort_field=self.input_sort_field, raw_variables=self.raw_variables, input_nets=self.input_nets)
```
Comment (samvanstroud) [inline src/hepattn/models/maskformer.py:121]: We definitely shouldn't be instantiating a new instance on every step, can you move this to the `__init__`

File: src/hepattn/callbacks/attn_mask_logger.py [hunk 15]
```diff
@@ -1,46 +1,162 @@
...
+
+                # Log stats if enabled
+                if self.log_stats:
+                    self._log_attention_stats(pl_module, mask, step, layer, f"{prefix_base}_strided")
+
             # Clear after logging
-            delattr(model, "attn_masks_to_log")
+            delattr(model.decoder, "strided_masks_to_log")
+
+    def sort_mask_by_phi(self, attn_mask_im, query_phi, key_phi):
```
Comment (samvanstroud) [inline src/hepattn/callbacks/attn_mask_logger.py:145]: Do we need the sorting here? presumably the mask is already presorted now?

File: src/hepattn/models/decoder.py [hunk 16]
```diff
@@ -45,13 +51,15 @@ def __init__(
...
-        self.tasks = None  # Will be set by MaskFormer
+        self.tasks: list | None = None  # Will be set by MaskFormer
         self.num_queries = num_queries
         self.mask_attention = mask_attention
         self.use_query_masks = use_query_masks
         self.log_attn_mask = log_attn_mask
+        self.key_posenc = key_posenc
         self.query_posenc = query_posenc
         self.preserve_posenc = preserve_posenc
+        self.posenc_analysis = posenc_analysis
```
Comment (samvanstroud) [inline src/hepattn/models/decoder.py:62]: What is this for? I saw `posenc_analysis_logging()` but not sure we still need it?

File: src/hepattn/models/decoder.py [hunk 17]
```diff
@@ -113,49 +131,64 @@ def forward(self, x: dict[str, Tensor], input_names: list[str]) -> tuple[dict[st
...
+
+    def posenc_analysis_logging(self, x, query_posenc, key_posenc):
+        if query_posenc is not None:
+            self.last_query_posenc = query_posenc.detach().cpu().numpy()
+            self.last_query_phi = x["query_phi"].detach().cpu().numpy()
+        if key_posenc is not None:
+            key_phi = x["key_phi"].detach().cpu().numpy()
+            self.last_key_phi = key_phi
+            key_posenc = key_posenc[0].cpu()
+            self.last_key_posenc = key_posenc.numpy()
```
Comment (samvanstroud) [inline src/hepattn/models/decoder.py:191]: As above, I'm not sure this is needed anymore

File: src/hepattn/models/maskformer.py [hunk 18]
```diff
@@ -145,8 +160,26 @@ def forward(self, inputs: dict[str, Tensor]) -> dict[str, Tensor]:
...
+            attn_mask_im = attn_logits[0].detach().cpu().clone().sigmoid()
+            self.output_attn_masks_to_log["final"] = {
+                "mask": attn_mask_im >= threshold,
+                "probs": attn_mask_im,
+                "step": self.log_step,
+                "layer": "final",
+            }
+            # store phi vals for sorting in callback
+            self.output_attn_masks_to_log["final"]["query_phi"] = x.get("query_phi")
+            self.output_attn_masks_to_log["final"]["key_phi"] = x.get("key_phi")
```
Comment (samvanstroud) [inline src/hepattn/models/maskformer.py:182]: Not sure we need this, everything should be presorted?

File: src/hepattn/models/maskformer.py [hunk 19]
```diff
@@ -145,8 +160,26 @@ def forward(self, inputs: dict[str, Tensor]) -> dict[str, Tensor]:
...
+            if not hasattr(self, "output_attn_masks_to_log"):
+                self.output_attn_masks_to_log = {}
+            # sigmoid the attn mask to get the probability of the hit being attended to
+            attn_mask_im = attn_logits[0].detach().cpu().clone().sigmoid()
+            self.output_attn_masks_to_log["final"] = {
+                "mask": attn_mask_im >= threshold,
+                "probs": attn_mask_im,
+                "step": self.log_step,
+                "layer": "final",
+            }
```
Comment (samvanstroud) [inline src/hepattn/models/maskformer.py:179]: If the `attn_logits` are stored in `outputs`, then I think we should be able to move all fo this to the callback, which has access to `trainer.global_step`, so we can remove the step counter here, and then retrieve the `attn_logits` and turn them into the binary mask in the callback itself

**Review by samvanstroud — CHANGES_REQUESTED** [review]
Thanks @pippaduckett1, I did a pass and think there is a bit more clean up to do before merging this. Let me know if anything is unclear! thanks :)

File: src/hepattn/models/maskformer.py (hunk 13 above)
Comment (pippaduckett1) [reply src/hepattn/models/maskformer.py:261]: A lot of these were from the linter giving errors - I think I need to change the settings maybe to match the repo lint

**Review by samvanstroud — CHANGES_REQUESTED** [review]
Thanks for the updates @pippaduckett1, a few hopefully last things:

File: src/hepattn/callbacks/attn_mask_logger.py [hunk 20]
```diff
@@ -1,46 +1,191 @@
...
+        # For validation, log every time. For training, log every N steps
+        if not is_validation and step % self.log_every_n_steps != 0:
+            return
+
+        # Process final outputs
+        if "final" in outputs:
+            for task_name, task_outputs in outputs["final"].items():
+                # Look for ObjectHitMaskTask outputs (they end with "_logit")
+                for output_key, output_value in task_outputs.items():
+                    if output_key.endswith("_logit") and isinstance(output_value, torch.Tensor):
```
[the remark this answers is not in the document]
Comment (samvanstroud) [reply src/hepattn/callbacks/attn_mask_logger.py]: ah yes ok makes sense thanks - in that case no harm logging plots for both I guess (though I would sigmoid the logits to normalise them at least). 

I think in any case the code to log the thresholded mask versus the mask probs should be very similar, so I would harmonise the `_log_attention_mask` and `_log_attention_logits` as much as possible to avoid redundancy

[the remark this answers is not in the document]
Comment (samvanstroud) [reply src/hepattn/callbacks/attn_mask_logger.py]: I'm also happy if you want to keep it simple for now and just log the mask - up to you
